modules/stereo/src/descriptor.hpp

```diff
@@ -0,0 +1,435 @@
+						{
+							for (int ii = i - n2; ii <= i + n2; ii += 2)
+							{
+								int rrWidth = ii * width;
+								for (int jj = j - n2; jj <= j + n2; jj += 2)
+								{
+									if (ii != i || jj != j)
+									{
+										//compare center pixel with pixel at distance 2x
+										if (image1[(rrWidth + jj)] > image1[(rWidth + j)])
+										{
+											c = c + 1;
+										}
+										c = c * 2;
+									}
+									if (ii != i || jj != j)
+									{
+										//same for image 2
+										if (image2[(rrWidth + jj)] > image2[(rWidth + j)])
+										{
+											c2 = c2 + 1;
+										}
+										c2 = c2 * 2;
+									}
+								}
+							}
+						}
+						dst1[(rWidth + j)] = c;
+						dst2[(rWidth + j)] = c2;
+					}
+				}
+			}
+		};
+		//!class that implemented the census descriptor on single images
+		class singleImageCensus : public ParallelLoopBody
+		{
+		private:
+			uint8_t *image;
+			int *dst;
+			int n2, width, height, type;
+		public:
+			singleImageCensus(uint8_t * img1, int w, int h, int k2, int * distance1,const int t) :
+				image(img1), dst(distance1), n2(k2), width(w), height(h), type(t){}
+
+			virtual void operator()(const cv::Range &r) const {
+
+				for (int i = r.start; i <= r.end ; i++)
+				{
+					int rWidth = i * width;
+					for (int j = n2; j <= width - n2; j++)
+					{
+						if (type == CV_SSE_CENSUS)
+						{
+							//to do
+						}
+						else
+						{
+							int c = 0;
+							for (int ii = i - n2; ii <= i + n2; ii++)
+							{
+								int rrWidth = ii * width;
+								for (int jj = j - n2; jj <= j + n2; jj++)
+								{
+									if (ii != i || jj != j)
+									{
+										if (image[(rrWidth + jj)] > image[(rWidth + j)])
+										{
+											c = c + 1;
+										}
+										c = c * 2;
+									}
+								}
+							}
+							dst[(rWidth + j)] = c;
+						}
+					}
+				}
+			}
+
+		};
+		//! parallel implementation of MCT type of descriptors
+		class parallelMctDescriptor:public ParallelLoopBody
+		{
+		private:
+			uint8_t *image1, *image2;
+			int *dst1, *dst2;
+			int n2,t , width, height, type;
+		public:
+			parallelMctDescriptor(uint8_t * img1, uint8_t * img2, int w, int h, int k2,int threshold, int * distance1, int * distance2,const int tip) :
+				image1(img1), image2(img2), dst1(distance1), dst2(distance2), n2(k2), t(threshold), width(w), height(h), type(tip){}
+
+			virtual void operator()(const cv::Range &r) const {
+
+				for (int i = r.start; i <= r.end ; i++)
+				{
+					int rWidth = i * width;
+					int distV = (i)* width;
+					for (int j = n2 + 2; j <= width - n2 - 2; j++)
+					{
+						int c = 0;
+						int c2 = 0;
+						if (type == CV_STANDARD_MCT)
+						{
+							for (int ii = i - n2; ii <= i + n2; ii += 2)
+							{
+								int rrWidth = ii * width;
+								for (int jj = j - n2; jj <= j + n2; jj += 2)
+								{
+									if (ii != i || jj != j)
+									{
+										if (image1[rrWidth + jj] > image1[rWidth + j] - t)
```

Comment:
I wonder if we can reduce the code duplication even further. Like, do you notice a significant speed penality if you place a `type switch` here? I believe that if it's under 10%, we can go for it. At least for now - the code is not optimized yet anyways.